### ISSUE TITLE
Typescript error with async factories

### ISSUE THREAD

**Issue by Ackos95** [jeffijoe/awilix#83]
I've been trying to create `async` app initialization using `awilix`. As `container` doesn't support `async` factories I've used pattern recommended [here](https://github.com/jeffijoe/awilix/issues/12#issuecomment-289419380), but typescript trows an error if I try to destruct dependencies in factory function declaration.

Example:
```js
const createContainer = async (): AwilixContainer => {
  const container: AwilixContainer = createContainer();

  const config: TConfigOptions = { dbOptions: { /* ... */ } };
  container.register('config', asValue(config));

  // passing cradle to ensure dependencies inside factory
  const db = await asyncDbFactory(container.cradle);
  container.register('db', asValue(db));
  // rest of the code
}

// db factory
// For this destruction I receive error for type missmatch
// `{ [key: string]: any }` doesn't match `{ config: TConfigOptions }`
const asyncDbFactory = ({ config }: { config: TConfigOptions }) =>
  new Promise((resolve: (db: TDb) => void) => {
    // connection logic

    // resolve(db);
  })
```

It does work properly if I don't define types in `asyncDbFactory`
```js
// this way typescript won't mind
const asyncDbFactory(opts: any) => /* ... */
```

Is there a way to bypass this that I'm not seeing or is it just a price we must pay?

**Comment by Ackos95** [comment]
If I may suggest simple solution. It can be done by modifying type of `container.cradle` into `any`, I know that sounds counter-intuitive but it would enable to manually define type of dependencies and to cast `cradle` into it when invoking. In previous example it would look something like this:

```js
const createContainer = async (): AwilixContainer => {
  const container: AwilixContainer = createContainer();

  const config: TConfigOptions = { dbOptions: { /* ... */ } };
  container.register('config', asValue(config));

  // passing cradle to ensure dependencies inside factory
  const db = await asyncDbFactory(container.cradle as { config: TConfigOptions });
  container.register('db', asValue(db));
  // rest of the code
}

// db factory
// No error anymore
const asyncDbFactory = ({ config }: { config: TConfigOptions }) =>
  new Promise((resolve: (db: TDb) => void) => {
    // connection logic

    // resolve(db);
  })
```

I'm aware that this is not a perfect solution, but current `cradle` type doesn't give much information, and yet it blocks any casting.

**Comment by jeffijoe** [comment]
I'm fine with that. 😄

As a tip, you can actually use a double cast to get around this:

```ts
container.cradle as any as { config: TConfigOptions }
```

I'll merge your PR and release.

**Comment by jeffijoe** [comment]
Released as v3.0.6, thanks!
